utt-replica/signature-processor/src/sigProcessor.cpp

```diff
@@ -0,0 +1,157 @@
+  };
+#pragma pack(pop)
+  static_assert(sizeof(Header) == (6 + 2 + 8 + 2), "Header is 18B");
+
+  Header* b() const { return (Header*)msgBody_; }
+};
+}  // namespace messages
+SigProcessor::SigProcessor(uint16_t repId,
+                           uint16_t n,
+                           uint16_t threshold,
+                           uint64_t timer_handler_timeout,
+                           std::shared_ptr<bftEngine::impl::MsgsCommunicator> msgs_communicator,
+                           std::shared_ptr<bftEngine::impl::MsgHandlersRegistrator> msg_handlers,
+                           concordUtil::Timers& timers)
+    : msgs_communicator_{msgs_communicator},
+      timers_{timers},
+      repId_{repId},
+      n_{n},
+      threshold_{threshold},
+      timer_handler_timeout_{timer_handler_timeout} {
+  msg_handlers->registerMsgHandler(bftEngine::impl::MsgCode::Reserved, [&](bftEngine::impl::MessageBase* message) {
+    messages::PartialSigMsg* msg = (messages::PartialSigMsg*)message;
+    onReceivingNewPartialSig(msg->getsig_id(), msg->idOfGeneratedReplica(), msg->getPartialSig());
+    delete msg;
+  });
+  timeout_handler_ = timers_.add(std::chrono::milliseconds(timer_handler_timeout_),
+                                 concordUtil::Timers::Timer::RECURRING,
+                                 [&](concordUtil::Timers::Handle h) {
+                                   (void)h;
+                                   const auto& entries = jobs_.getEntries();
+                                   for (const auto& [id, entry] : entries) {
+                                     (void)id;
+                                     if (entry.start_time == 0) continue;
+                                     auto now = std::chrono::steady_clock::now().time_since_epoch().count();
+                                     if (now - entry.start_time >= entry.job_timeout_ms) {
+                                       onJobTimeout(entry);
+                                     }
+                                   }
+                                 });
+}
+
+void SigProcessor::processSignature(uint64_t sig_id,
+                                    const std::vector<uint8_t>& sig,
+                                    const validationCb& vcb,
+                                    uint64_t job_timeout_ms) {
+  jobs_.registerEntry(sig_id, vcb, job_timeout_ms);
+  const auto& jobEntry = jobs_.get(sig_id);
+  auto& validator = jobs_.getJobValidator(sig_id);
+  std::vector<uint16_t> invalids;
+  for (auto& [id, psig] : jobEntry.partial_sigs) {
+    if (!validator(id, psig)) invalids.push_back((uint16_t)id);
+  }
+  for (auto id : invalids) jobs_.removePartialSig(sig_id, (uint32_t)id);
+
+  jobs_.addPartialSig(sig_id, (uint32_t)repId_, sig);
+  auto job_coordinator = get_job_coordinator_cb(sig_id);
+  if (job_coordinator == repId_) return;
+  messages::PartialSigMsg msg(repId_, sig, sig_id);
+  msgs_communicator_->sendAsyncMessage((uint64_t)job_coordinator, msg.body(), msg.size());
```

Comment:
This is a bit of mess. You get a jobEntry const ref that you read from but at the same time you modify it multiple times through other SigJobEntriesMap methods. Maybe if you try to get rid of SigJobEntriesMap and do the locking in sig processor you will end up with much simpler and straightforward code.

Reply:
The main reason for introducing the SigJobEntriesMap is to have the sigProcessor code focused only on the sigProcessor logic (while introducing a fine granularity locking as possible).
As for the specific comment, we don't change the entry while reading it, we only get the const reference and validate the signatures outside of the lock (while letting other threads add/remove data to the complete job maps).  

of course, all of this can be done inside the code of the sigPorcessor, but as I said, I personally find it easier to focus on the signature collecting process only here and leave the storage/locking and all the dirty things to some another internal component

Reply:
Basically, the challenge here is not to block the other threads while validating the partial signatures.
Now, recall that SigJobEntriesMap is not supposed to be a complete thread-safe data structure, but more of a specific implementation of thread-safe container for the signature collector.
For example, we do know that an honest replica would never change the entry after it was registered.
We also know who are the callers of the methods (the signer threads and the message processing thread).
I'm not sure it worth investing in building a complete thread-safe data structure, and locking in the sigProcessor level, will probably introduce similar issues if we want to keep the fine granularity of the locking... 
I will try to design maybe a more comprehensible data container layer.
Anyway, regarding the comment itself, we don't change the entry while reading it, we first read it (and let other threads to grab the lock), then we accumulate the invalid signatures, then we grab the lock again and remove the invalid signatures.
Another maybe straightforward solution is to add a mutex per entry, but again, I think that considering the involved threads and the data flow, it might quite overkill.
WDYT?